File: backend-php/src/api/handlers/query.php
<?php

namespace Api\Handlers;

use Db\Database;
use Db\Models\GameUtils;

class QueryHandler {
    private $pdo;
    private $notificationHandler;

    public function __construct($notificationHandler = null) {
        $this->pdo = Database::getInstance();
        $this->notificationHandler = $notificationHandler;
    }

    /**
     * Create query
     */
    public function createQuery($data) {
        try {
            $patientId = $_SERVER['userID'] ?? null;
            if (!$patientId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            if (empty($data['question'])) {
                return [
                    'status' => 400,
                    'response' => ['error' => 'Question is required']
                ];
            }

            // Verify patient exists
            $stmt = $this->pdo->prepare('SELECT * FROM users WHERE id = ? AND role = ?');
            $stmt->execute([$patientId, 'patient']);
            $patient = $stmt->fetch(\PDO::FETCH_ASSOC);

            if (!$patient) {
                return [
                    'status' => 403,
                    'response' => ['error' => 'Only patients can create queries']
                ];
            }

            $now = date('Y-m-d H:i:s');
            $urgency = $data['urgency'] ?? 'medium';
            $status = 'pending';

            // If patient has a doctor assigned, set the doctorId
            $doctorId = null;
            if (!empty($patient['doctor_id'])) {
                // Look up doctor by UUID
                $stmt = $this->pdo->prepare('SELECT id FROM users WHERE id = ?');
                $stmt->execute([$patient['doctor_id']]);
                $doctorRow = $stmt->fetch(\PDO::FETCH_ASSOC);
                if ($doctorRow) {
                    $doctorId = $doctorRow['id'];
                } else {
                    return [
                        'status' => 400,
                        'response' => ['error' => 'U Need a Valid doctor ID']
                    ];
                }
            } else {
                return [
                    'status' => 400,
                    'response' => ['error' => 'U Need a doctor ID']
                ];
            }

            $stmt = $this->pdo->prepare('
                INSERT INTO queries (patientId, doctorId, question, status, urgency, createdAt, updatedAt)
                VALUES (?, ?, ?, ?, ?, ?, ?)
            ');

            $stmt->execute([$patientId, $doctorId, $data['question'], $status, $urgency, $now, $now]);
            $queryId = $this->pdo->lastInsertId();

            // Return created query
            $stmt = $this->pdo->prepare('SELECT * FROM queries WHERE id = ?');
            $stmt->execute([$queryId]);
            $query = $stmt->fetch(\PDO::FETCH_ASSOC);

            return [
                'status' => 201,
                'response' => ['query' => $query]
            ];
        } catch (\Exception $e) {
            return [
                'status' => 500,
                'response' => ['error' => 'Database error']
            ];
        }
    }

    /**
     * Get query by ID
     */
    public function getQuery($queryId) {
        try {
            $userId = $_SERVER['userID'] ?? null;
            if (!$userId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            // Get user role
            $stmt = $this->pdo->prepare('SELECT role FROM users WHERE id = ?');
            $stmt->execute([$userId]);
            $userRow = $stmt->fetch(\PDO::FETCH_ASSOC);

            if (!$userRow) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User not found']
                ];
            }

            $userRole = $userRow['role'];

            // Get the query
            $stmt = $this->pdo->prepare('
                SELECT q.*, u1.fullName as patientName, u2.fullName as doctorName
                FROM queries q
                LEFT JOIN users u1 ON q.patientId = u1.id
                LEFT JOIN users u2 ON q.doctorId = u2.id
                WHERE q.id = ?
            ');
            $stmt->execute([$queryId]);
            $query = $stmt->fetch(\PDO::FETCH_ASSOC);

            if (!$query) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Query not found']
                ];
            }

            // Verify access: patient can only see their own queries, doctor can see assigned queries
            if ($userRole === 'patient' && $query['patientId'] != $userId) {
                return [
                    'status' => 403,
                    'response' => ['error' => 'Forbidden']
                ];
            } elseif ($userRole === 'doctor' && $query['doctorId'] != $userId) {
                return [
                    'status' => 403,
                    'response' => ['error' => 'Forbidden']
                ];
            }

            // Get full patient object
            $patient = null;
            if ($query['patientId']) {
                $stmt = $this->pdo->prepare('SELECT * FROM users WHERE id = ?');
                $stmt->execute([$query['patientId']]);
                $patient = $stmt->fetch(\PDO::FETCH_ASSOC);
                if ($patient) {
                    unset($patient['password']);
                }
            }

            // Get full doctor object
            $doctor = null;
            if ($query['doctorId']) {
                $stmt = $this->pdo->prepare('SELECT * FROM users WHERE id = ?');
                $stmt->execute([$query['doctorId']]);
                $doctor = $stmt->fetch(\PDO::FETCH_ASSOC);
                if ($doctor) {
                    unset($doctor['password']);
                }
            }

            return [
                'status' => 200,
                'response' => [
                    'query' => $query,
                    'patient' => $patient,
                    'doctor' => $doctor
                ]
            ];
        } catch (\Exception $e) {
            error_log('Error in getQuery: ' . $e->getMessage());
            return [
                'status' => 500,
                'response' => ['error' => 'Database error']
            ];
        }
    }

    /**
     * Answer query (doctor only)
     */
    public function answerQuery($queryId, $data) {
        try {
            $doctorId = $_SERVER['userID'] ?? null;
            if (!$doctorId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            if (empty($data['response'])) {
                return [
                    'status' => 400,
                    'response' => ['error' => 'Response is required']
                ];
            }

            // Get query
            $stmt = $this->pdo->prepare('SELECT * FROM queries WHERE id = ?');
            $stmt->execute([$queryId]);
            $query = $stmt->fetch(\PDO::FETCH_ASSOC);

            if (!$query) {
                return [
                    'status' => 404,
                    'response' => ['error' => 'Query not found']
                ];
            }

            // Verify doctor is assigned to this query or can assign themselves
            if ($query['doctorId'] && $query['doctorId'] != $doctorId) {
                return [
                    'status' => 403,
                    'response' => ['error' => 'This query is assigned to another doctor']
                ];
            }

            // Update query
            $now = date('Y-m-d H:i:s');
            $stmt = $this->pdo->prepare('
                UPDATE queries
                SET doctorId = ?, response = ?, status = ?, answeredAt = ?, updatedAt = ?
                WHERE id = ?
            ');

            $stmt->execute([$doctorId, $data['response'], 'answered', $now, $now, $queryId]);

            return [
                'status' => 200,
                'response' => ['message' => 'Query answered successfully']
            ];
        } catch (\Exception $e) {
            return [
                'status' => 500,
                'response' => ['error' => 'Database error']
            ];
        }
    }

    /**
     * Get all queries (for doctors)
     */
    public function getAllQueries() {
        try {
            $doctorId = $_SERVER['userID'] ?? null;
            if (!$doctorId) {
                return [
                    'status' => 401,
                    'response' => ['error' => 'User ID not found']
                ];
            }

            // Get query parameters
            $status = $_GET['status'] ?? null;
            $includeAll = $_GET['includeAll'] ?? 'false';
            $includeAll = strtolower($includeAll) === 'true';

            // Build WHERE clause
            $where = 'WHERE q.doctorId = ?';
            $params = [$doctorId];

            // Add status filter if provided
            if ($status && $status !== '') {
                $where .= ' AND q.status = ?';
                $params[] = $status;
            }

            $sql = '
                SELECT q.*, u1.fullName as patientName, u2.fullName as doctorName
                FROM queries q
                LEFT JOIN users u1 ON q.patientId = u1.id
                LEFT JOIN users u2 ON q.doctorId = u2.id
                ' . $where . '
                ORDER BY 
                    CASE q.urgency
                        WHEN "high" THEN 1
                        WHEN "medium" THEN 2
                        WHEN "low" THEN 3
                    END,
                    q.createdAt DESC
            ';

            $stmt = $this->pdo->prepare($sql);
            $stmt->execute($params);
            $queries = $stmt->fetchAll(\PDO::FETCH_ASSOC);

            return [
                'status' => 200,
                'response' => ['queries' => $queries]
            ];
        } catch (\Exception $e) {
            error_log('Error in getAllQueries: ' . $e->getMessage());
            return [
                'status' => 500,
                'response' => ['error' => 'Database error']
            ];
        }
    }
}

File: backend/internal/api/handlers/query.go
package handlers

import (
	"context"
	"net/http"
	"time"

	"github.com/Meheer17/ambieye/internal/db/models"
	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// QueryHandler handles query-related operations
type QueryHandler struct {
	queryCollection        *mongo.Collection
	userCollection         *mongo.Collection
	notificationCollection *mongo.Collection
	notificationHandler    *NotificationHandler
}

// NewQueryHandler creates a new QueryHandler
func NewQueryHandler(db *mongo.Database, notificationHandler *NotificationHandler) *QueryHandler {
	return &QueryHandler{
		queryCollection:        db.Collection("queries"),
		userCollection:         db.Collection("users"),
		notificationCollection: db.Collection("notifications"),
		notificationHandler:    notificationHandler,
	}
}

// CreateQuery handles the creation of a new query by a patient
func (h *QueryHandler) CreateQuery(c *gin.Context) {
	patientID, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
		return
	}

	var queryData struct {
		Question string `json:"question" binding:"required"`
		Urgency  string `json:"urgency"`
	}

	if err := c.ShouldBindJSON(&queryData); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body" + err.Error()})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Verify that the patient exists
	var patient models.User
	err := h.userCollection.FindOne(ctx, bson.M{
		"_id":  patientID,
		"role": "patient",
	}).Decode(&patient)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusForbidden, gin.H{"error": "Only patients can create queries"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Create the query object
	now := time.Now()
	query := models.Query{
		ID:        primitive.NewObjectID(),
		PatientID: patientID.(primitive.ObjectID),
		Question:  queryData.Question,
		Status:    "pending",
		CreatedAt: now,
		UpdatedAt: now,
	}

	// If patient has an assigned doctor, set the query's doctorID
	if patient.DoctorID != "" {
		var doctor models.User
		err = h.userCollection.FindOne(ctx, bson.M{"uuid": patient.DoctorID}).Decode(&doctor)
		if err == nil {
			query.DoctorID = doctor.ID
		} else {
			c.JSON(http.StatusBadRequest, gin.H{"error": "U Need a Valid doctor ID"})

		}
	} else {
		c.JSON(http.StatusBadRequest, gin.H{"error": "U Need a doctor ID"})
	}

	// Set urgency if provided
	if queryData.Urgency != "" {
		query.Urgency = queryData.Urgency
	} else {
		query.Urgency = "medium" // Default urgency
	}

	// var doctorID primitive.ObjectID
	// Insert the query into the database
	_, err = h.queryCollection.InsertOne(ctx, query)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create query"})
		return
	}

	// Create notification for doctor if one was assigned
	// if !doctorID.IsZero() {
	// 	notificationData := map[string]string{
	// 		"queryId":     query.ID.Hex(),
	// 		"patientId":   patientID.(primitive.ObjectID).Hex(),
	// 		"patientName": patient.FullName,
	// 	}

	// 	// Create notification for the doctor
	// 	_, err = h.notificationHandler.CreateNotification(
	// 		ctx,
	// 		doctorID,
	// 		models.NotificationTypeNewQuery,
	// 		"New Query",
	// 		"You have received a new query from "+patient.FullName,
	// 		notificationData,
	// 	)

	// 	if err != nil {
	// 		// Log error but don't fail the request
	// 		c.Error(err)
	// 	}
	// }

	c.JSON(http.StatusCreated, gin.H{"query": query})
}

// GetQuery retrieves a specific query by ID
func (h *QueryHandler) GetQuery(c *gin.Context) {
	userID, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
		return
	}

	role, exists := c.Get("role")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User role not found"})
		return
	}

	queryID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid query ID"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Find the query
	var query models.Query
	err = h.queryCollection.FindOne(ctx, bson.M{"_id": queryID}).Decode(&query)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"error": "Query not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Verify that the user has access to this query (either as the patient or as the assigned doctor)
	if role == "patient" && query.PatientID != userID {
		c.JSON(http.StatusForbidden, gin.H{"error": "You don't have access to this query"})
		return
	}

	if role == "doctor" && !query.DoctorID.IsZero() && query.DoctorID != userID {
		c.JSON(http.StatusForbidden, gin.H{"error": "You don't have access to this query"})
		return
	}

	// Get patient information
	var patient models.User
	err = h.userCollection.FindOne(ctx, bson.M{"_id": query.PatientID}).Decode(&patient)
	if err != nil && err != mongo.ErrNoDocuments {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Get doctor information if available
	var doctor models.User
	if !query.DoctorID.IsZero() {
		err = h.userCollection.FindOne(ctx, bson.M{"_id": query.DoctorID}).Decode(&doctor)
		if err != nil && err != mongo.ErrNoDocuments {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
			return
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"query":   query,
		"patient": patient.ToSafeUser(),
		"doctor":  doctor.ToSafeUser(),
	})
}

// AnswerQuery allows a doctor to answer a patient's query
func (h *QueryHandler) AnswerQuery(c *gin.Context) {
	doctorID, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
		return
	}

	role, exists := c.Get("role")
	if !exists || role != "doctor" {
		c.JSON(http.StatusForbidden, gin.H{"error": "Only doctors can answer queries"})
		return
	}

	queryID, err := primitive.ObjectIDFromHex(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid query ID"})
		return
	}

	var answerData struct {
		Response string `json:"response" binding:"required"`
	}

	if err := c.ShouldBindJSON(&answerData); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Find the query
	var query models.Query
	err = h.queryCollection.FindOne(ctx, bson.M{"_id": queryID}).Decode(&query)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			c.JSON(http.StatusNotFound, gin.H{"error": "Query not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// If query is already answered, don't allow changes
	if query.Status == "answered" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Query has already been answered"})
		return
	}

	// If the query is assigned to a specific doctor, verify it's this doctor
	if !query.DoctorID.IsZero() && query.DoctorID != doctorID {
		c.JSON(http.StatusForbidden, gin.H{"error": "This query is assigned to another doctor"})
		return
	}

	// Update the query with the answer
	now := time.Now()
	update := bson.M{
		"$set": bson.M{
			"response":   answerData.Response,
			"status":     "answered",
			"doctorId":   doctorID,
			"updatedAt":  now,
			"answeredAt": now,
		},
	}

	_, err = h.queryCollection.UpdateOne(ctx, bson.M{"_id": queryID}, update)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update query"})
		return
	}

	// Get doctor information for notification
	var doctor models.User
	err = h.userCollection.FindOne(ctx, bson.M{"_id": doctorID}).Decode(&doctor)
	if err != nil {
		// Log the error but don't fail the request
		c.Error(err)
	}

	// Create notification for the patient
	notificationData := map[string]string{
		"queryId":    queryID.Hex(),
		"doctorId":   doctorID.(primitive.ObjectID).Hex(),
		"doctorName": doctor.FullName,
	}

	_, err = h.notificationHandler.CreateNotification(
		ctx,
		query.PatientID,
		models.NotificationTypeQueryAnswer,
		"Query Answered",
		"Your query has been answered by Dr. "+doctor.FullName,
		notificationData,
	)

	if err != nil {
		// Log error but don't fail the request
		c.Error(err)
	}

	c.JSON(http.StatusOK, gin.H{"message": "Query answered successfully"})
}

func (h *QueryHandler) GetAllQueries(c *gin.Context) {
	doctorID, exists := c.Get("userID")
	if !exists {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User ID not found"})
		return
	}

	role, exists := c.Get("role")
	if !exists || role != "doctor" {
		c.JSON(http.StatusForbidden, gin.H{"error": "Only doctors can access all queries"})
		return
	}

	status := c.Query("status") // Optional filter by status

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Build filter
	filter := bson.M{}
	includeAll := c.Query("includeAll")
	// Check if includeAll parameter exists and is set to "true"
	if includeAll == "true" {
		// If includeAll is true, don't filter by doctor ID
		// But if status is provided, still filter by status
		if status != "" {
			filter["status"] = status
		}
	} else {
		// Normal case - filter by doctor ID
		filter["doctorId"] = doctorID
		if status != "" {
			filter["status"] = status
		}
	}

	// Get queries with pagination
	page, perPage := getPaginationParams(c)
	opts := options.Find().
		SetSort(bson.M{"createdAt": -1}).
		SetSkip(int64((page - 1) * perPage)).
		SetLimit(int64(perPage))

	cursor, err := h.queryCollection.Find(ctx, filter, opts)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}
	defer cursor.Close(ctx)

	var queries []models.Query
	if err := cursor.All(ctx, &queries); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Get total count for pagination
	totalCount, err := h.queryCollection.CountDocuments(ctx, filter)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Database error"})
		return
	}

	// Prepare enhanced queries with patient information
	queriesWithInfo := make([]models.QueryWithUserInfo, 0, len(queries))
	for _, query := range queries {
		// Get patient information
		var patient models.User
		err := h.userCollection.FindOne(ctx, bson.M{"_id": query.PatientID}).Decode(&patient)
		if err == nil { // Only add if patient found
			queryWithInfo := models.QueryWithUserInfo{
				ID:          query.ID,
				PatientID:   query.PatientID,
				PatientName: patient.FullName,
				DoctorID:    query.DoctorID,
				Question:    query.Question,
				Response:    query.Response,
				Status:      query.Status,
				Urgency:     query.Urgency,
				CreatedAt:   query.CreatedAt,
				UpdatedAt:   query.UpdatedAt,
				AnsweredAt:  query.AnsweredAt,
			}
			queriesWithInfo = append(queriesWithInfo, queryWithInfo)
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"queries": queriesWithInfo,
		"pagination": gin.H{
			"page":       page,
			"perPage":    perPage,
			"totalItems": totalCount,
			"totalPages": (totalCount + int64(perPage) - 1) / int64(perPage),
		},
	})
}
